tachyon/c/zk/plonk/halo2/bn254_prover.cc

```diff
@@ -177,6 +177,58 @@ tachyon_bn254_g1_projective* CommitLagrange(
   return prover->CommitLagrangeRaw(scalars);
 }
 
+template <typename NativeProver>
+void BatchStart(NativeProver* prover, size_t len) {
+  using PCS = typename NativeProver::PCS;
+
+  if constexpr (PCS::kSupportsBatchMode) {
+    prover->pcs().SetBatchMode(len);
+  } else {
+    NOTREACHED() << "PCS doesn't support batch commitment";
+  }
+}
+
+template <typename NativeProver>
+void BatchCommit(NativeProver* prover,
+                 const tachyon_bn254_univariate_dense_polynomial* poly,
+                 size_t idx) {
+  using PCS = typename NativeProver::PCS;
+
+  if constexpr (PCS::kSupportsBatchMode) {
+    prover->BatchCommitAt(c::base::native_cast(*poly), idx);
+  } else {
+    NOTREACHED() << "PCS doesn't support batch commitment";
+  }
+}
+
+template <typename NativeProver>
+void BatchCommitLagrange(NativeProver* prover,
+                         const tachyon_bn254_univariate_evaluations* evals,
+                         size_t idx) {
+  using PCS = typename NativeProver::PCS;
+
+  if constexpr (PCS::kSupportsBatchMode) {
+    prover->BatchCommitAt(c::base::native_cast(*evals), idx);
+  } else {
+    NOTREACHED() << "PCS doesn't support batch commitment";
+  }
+}
+
+template <typename NativeProver>
+void BatchEnd(NativeProver* prover, tachyon_bn254_g1_affine* points,
+              size_t len) {
+  using PCS = typename NativeProver::PCS;
+  using Commitment = typename PCS::Commitment;
+
+  if constexpr (PCS::kSupportsBatchMode) {
+    std::vector<Commitment> commitments = prover->pcs().GetBatchCommitments();
+    CHECK_EQ(commitments.size(), len);
+    memcpy(points, commitments.data(), len * sizeof(Commitment));
```

Comment:
Is this memcpy avoidable? If it's too big change then maybe we can leave a `TODO` at least.

Reply:
Good point! I think it's avoidable, but it requires changes at `GetBatchCommitments()`! I'll leave the comment!